numpy/lib/function_base.py

```diff
@@ -3637,80 +3875,157 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
     if axis is None:
         axis = 0
 
-    Nx = ap.shape[axis]
-    indices = q * (Nx - 1)
+    # move axis to -1 for np.vectorize() operations.  Will move back later.
+    ap = np.swapaxes(ap, axis, -1)
+
+    if weights is None:
+        Nx = ap.shape[-1]
+        indices = q * (Nx - 1)
+    else:
+        # need a copy of weights for later array assignment
+        weights = np.swapaxes(weights.astype('f8'), axis, -1)
+        weights[weights < 0.] = 0.  # negative weights are treated as 0
+        # values with weight=0 are assigned minimum value and later moved left
+        abs_min = np.amin(ap)
+        ap[weights == 0.] = abs_min - 1.
+
+        def _sort_by_index(vector, vec_indices):
+            return vector[vec_indices]
+        # this func vectorizes sort along axis
+        arraysort = np.vectorize(_sort_by_index, signature='(i),(i)->(i)')
+
+        ind_sorted = np.argsort(ap, axis=-1)  # sort values long axis
+        ap_sorted = arraysort(ap, ind_sorted).astype('f8')
+
+        n = np.isnan(ap_sorted[..., -1:])
+        if n.ndim > 1:
+            n = np.swapaxes(n, axis, -1)
+
+        ws_sorted = arraysort(weights, ind_sorted).astype('f8')
+        ws_sorted[np.isnan(ap_sorted)] = 0.  # neglect nans from calculation
+        nonzero_w_inds = ws_sorted > 0.
+
+        cum_w = ws_sorted.cumsum(axis=-1)
+        cum_w_max = cum_w.max(axis=-1)
+
+        # some manipulation to get lower/upper percentage bounds
+        normalized_w_upper = (cum_w.T / cum_w_max.T).T
```

Comment:
Transpose seems pointless

Reply:
Imagine a case where
```
>>> ar = np.arange(24).reshape(2, 3, 4)
>>> ar
array([[[ 0,  1,  2,  3],
             [ 4,  5,  6,  7],
             [ 8,  9, 10, 11]],

           [[12, 13, 14, 15],
             [16, 17, 18, 19],
             [20, 21, 22, 23]]])
>>> axis = 1
>>> weights = [0, 1, 2]
>>> np.quantile(ar, q=0.5, axis=axis, weights=weights)
```
Then the calls to `np.swapaxes()` on lines 3923 and 3930 lead to
```
(Pdb) ap
array([[[-1,  4,  8],
             [-1,  5,  9],
             [-1,  6, 10],
             [-1,  7, 11]],

            [[-1, 16, 20],
              [-1, 17, 21],
              [-1, 18, 22],
              [-1, 19, 23]]])
(Pdb) ap.shape
(2, 4, 3)
(Pdb) weights
array([[[0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.]],

           [[0., 1., 2.],
            [0., 1., 2.],
            [0., 1., 2.],
            [0., 1., 2.]]])
(Pdb) weights.shape
(2, 4, 3)
```
Of course, `ap_sorted` and `ws_sorted` would have the same shapes:
```
(Pdb) ws_sorted
array([[[0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.]],

            [[0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.]]])
(Pdb) ap_sorted
array([[[-1.,  4.,  8.],
             [-1.,  5.,  9.],
             [-1.,  6., 10.],
             [-1.,  7., 11.]],

            [[-1., 16., 20.],
             [-1., 17., 21.],
             [-1., 18., 22.],
             [-1., 19., 23.]]])
```
Then, `cum_w` would have shape (2, 4, 3), and `cum_w_max` (2, 4):
```
(Pdb) cum_w
array([[[0., 1., 3.],
             [0., 1., 3.],
             [0., 1., 3.],
             [0., 1., 3.]],

           [[0., 1., 3.],
             [0., 1., 3.],
             [0., 1., 3.],
             [0., 1., 3.]]])
(Pdb) cum_w.shape
(2, 4, 3)
(Pdb) cum_w_max
array([[3., 3., 3., 3.],
            [3., 3., 3., 3.]])
(Pdb) cum_w_max.shape
(2, 4)
```
At this point, I think the only way to divide `cum_w` by `cum_w_max` and get back a shape (2, 4, 3) is to do 
```
(Pdb) (cum_w.T / cum_w_max.T).T
array([[[0.        , 0.33333333, 1.        ],
             [0.        , 0.33333333, 1.        ],
             [0.        , 0.33333333, 1.        ],
             [0.        , 0.33333333, 1.        ]],

           [[0.        , 0.33333333, 1.        ],
            [0.        , 0.33333333, 1.        ],
            [0.        , 0.33333333, 1.        ],
            [0.        , 0.33333333, 1.        ]]])
(Pdb) (cum_w.T / cum_w_max.T).T.shape
(2, 4, 3)
```

Reply:
Ah ok, I see what you're trying to do here. Can you just use `max(keepdims=True)` to keep the result broadcastable?

Reply:
But what if my input weights is the same shape as the data array, with different cell values?
```
>>> weights = np.arange(24).reshape(2,3,4)
```
Then what I want is
```
(Pdb) cum_w_max
array([[12., 15., 18., 21.],
       [48., 51., 54., 57.]])
```
But `.max(keepdims=True)` or `.max(axis=-1, keepdims=True)` would instead give me
```
(Pdb) cum_w.max(keepdims=True)
array([[[57.]]])
(Pdb) cum_w.max(axis=-1, keepdims=True)
array([[[12.],
        [15.],
        [18.],
        [21.]],

       [[48.],
        [51.],
        [54.],
        [57.]]])
```

Reply:
Normalization along any axis is best spelt `a / a.max(axis=ax, keepdims=True)` - keepdims avoids the need to do any juggling of shape. Sorry for not being clearer before